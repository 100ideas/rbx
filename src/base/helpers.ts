import classNames from "classnames";
import * as PropTypes from "prop-types";

import { Omit, Prefer } from "../types";
import { tuple } from "../utils";

/**
 * Float
 * https://github.com/jgthms/bulma/blob/master/sass/base/helpers.sass
 */
export const FLOAT_PULLED_ALIGNMENTS = tuple("left", "right");
export type FloatPulledAlignments = (typeof FLOAT_PULLED_ALIGNMENTS)[number];

export type FloatHelpersProps = Partial<{
  /** Fixes an element's floating children */
  clearfix: boolean;
  /** Moves an element to the left or right */
  pull: FloatPulledAlignments;
}>;

export const floatHelpersPropTypes = {
  clearfix: PropTypes.bool,
  pull: PropTypes.oneOf(FLOAT_PULLED_ALIGNMENTS),
};

export const transformFloatHelpers: TransformFunc<FloatHelpersProps> = (
  props,
  componentName,
  location = "prop",
) => {
  PropTypes.checkPropTypes(
    floatHelpersPropTypes,
    props,
    location,
    componentName,
  );
  const { className, clearfix, pull, ...rest } = props;

  return {
    className: classNames(
      {
        "is-clearfix": clearfix,
        [`is-pulled-${pull}`]: pull,
      },
      className,
    ),
    ...rest,
  };
};

/**
 * Overflow
 * https://github.com/jgthms/bulma/blob/master/sass/base/helpers.sass
 */
export type OverflowHelpersProps = Partial<{
  /** Adds overflow hidden */
  clipped: boolean;
}>;

export const overflowHelpersPropTypes = {
  clipped: PropTypes.bool,
};

export const transformOverflowHelpers: TransformFunc<OverflowHelpersProps> = (
  props,
  componentName,
  location = "prop",
) => {
  PropTypes.checkPropTypes(
    overflowHelpersPropTypes,
    props,
    location,
    componentName,
  );
  const { className, clipped, ...rest } = props;

  return {
    className: classNames({ "is-clipped": clipped }, className),
    ...rest,
  };
};

/**
 * Overlay
 * https://github.com/jgthms/bulma/blob/master/sass/base/helpers.sass
 */
export type OverlayHelpersProps = Partial<{
  /** Completely covers the first positioned parent */
  overlay: boolean;
}>;

export const overlayHelpersPropTypes = {
  overlay: PropTypes.bool,
};

export const transformOverlayHelpers: TransformFunc<OverlayHelpersProps> = (
  props,
  componentName,
  location = "prop",
) => {
  PropTypes.checkPropTypes(
    overlayHelpersPropTypes,
    props,
    location,
    componentName,
  );
  const { className, overlay, ...rest } = props;

  return {
    className: classNames({ "is-overlay": overlay }, className),
    ...rest,
  };
};

/**
 * Typography
 * https://github.com/jgthms/bulma/blob/master/sass/base/helpers.sass
 */
export const COLORS = tuple(
  "primary",
  "success",
  "info",
  "warning",
  "danger",
  "light",
  "dark",
  "white",
  "black",
  "link",
);
export type Colors = (typeof COLORS)[number];

export const GREY_COLORS = tuple(
  "black-bis",
  "black-ter",
  "grey-darker",
  "grey-dark",
  "grey",
  "grey-light",
  "grey-lighter",
  "white-ter",
  "white-bis",
);
export type GreyColors = (typeof GREY_COLORS)[number];

export const TEXT_ALIGNMENTS = tuple("centered", "justified", "left", "right");
export type TextAlignments = (typeof TEXT_ALIGNMENTS)[number];

export const TEXT_SIZES = tuple(1, 2, 3, 4, 5, 6);
export type TextSizes = (typeof TEXT_SIZES)[number];

export const TEXT_TRANSFORMS = tuple("capitalized", "lowercase", "uppercase");
export type TextTransforms = (typeof TEXT_TRANSFORMS)[number];

export const TEXT_WEIGHTS = tuple("light", "normal", "semibold", "bold");
export type TextWeights = (typeof TEXT_WEIGHTS)[number];

export type TypographyHelpersProps = Partial<{
  backgroundColor: Colors | GreyColors;
  italic: boolean;
  textAlignment: TextAlignments;
  textColor: Colors | GreyColors;
  textSize: TextSizes;
  textTransform: TextTransforms;
  textWeight: TextWeights;
}>;

export const typographyHelpersPropTypes = {
  backgroundColor: PropTypes.oneOf([...COLORS, ...GREY_COLORS]),
  italic: PropTypes.bool,
  textAlignment: PropTypes.oneOf(TEXT_ALIGNMENTS),
  textColor: PropTypes.oneOf([...COLORS, ...GREY_COLORS]),
  textSize: PropTypes.oneOf(TEXT_SIZES),
  textTransform: PropTypes.oneOf(TEXT_TRANSFORMS),
  textWeight: PropTypes.oneOf(TEXT_WEIGHTS),
};

export const transformTypographyHelpers: TransformFunc<
  TypographyHelpersProps
> = (props, componentName, location = "prop") => {
  PropTypes.checkPropTypes(
    typographyHelpersPropTypes,
    props,
    location,
    componentName,
  );
  const {
    backgroundColor,
    className,
    italic,
    textAlignment,
    textColor,
    textSize,
    textTransform,
    textWeight,
    ...rest
  } = props;

  return {
    className: classNames(
      {
        [`has-background-${backgroundColor}`]: backgroundColor,
        [`has-text-${textColor}`]: textColor,
        "is-italic": italic,
        [`is-${textTransform}`]: textTransform,
        [`has-text-${textAlignment}`]: textAlignment,
        [`has-text-weight-${textWeight}`]: textWeight,
        [`is-size-${textSize}`]: textSize,
      },
      className,
    ),
    ...rest,
  };
};

/**
 * Visibility
 * https://github.com/jgthms/bulma/blob/master/sass/base/helpers.sass
 */
export const DISPLAYS = tuple(
  "block",
  "flex",
  "inline",
  "inline-block",
  "inline-flex",
);
export type Displays = (typeof DISPLAYS)[number];

export type VisibilityHelpersProps = Partial<{
  /** Hides an element (unclear on where this is documented in Bulma) */
  hidden: boolean;
  /** Adds visibility hidden */
  invisible: boolean;
  /**
   * Hide elements visually but keep the element available to be announced by a
   * screen reader
   */
  srOnly: boolean;
}>;

export const visibilityHelpersPropTypes = {
  hidden: PropTypes.bool,
  invisible: PropTypes.bool,
  srOnly: PropTypes.bool,
};

export const transformVisibilityHelpers: TransformFunc<
  VisibilityHelpersProps
> = (props, componentName, location = "prop") => {
  PropTypes.checkPropTypes(
    visibilityHelpersPropTypes,
    props,
    location,
    componentName,
  );
  const { className, hidden, invisible, srOnly, ...rest } = props;

  return {
    className: classNames(
      {
        "is-hidden": hidden,
        "is-invisible": invisible,
        "is-sr-only": srOnly,
      },
      className,
    ),
    ...rest,
  };
};

/**
 * Other
 * https://github.com/jgthms/bulma/blob/master/sass/base/helpers.sass
 */
export type OtherHelpersProps = Partial<{
  /** Removes any margin */
  marginless: boolean;
  /** Removes any padding */
  paddingless: boolean;
  /** Removes any radius */
  radiusless: boolean;
  /** Removes any shadow */
  shadowless: boolean;
  /** Prevents the text from being selectable */
  unselectable: boolean;
}>;

export const otherHelpersPropTypes = {
  marginless: PropTypes.bool,
  paddingless: PropTypes.bool,
  radiusless: PropTypes.bool,
  shadowless: PropTypes.bool,
  unselectable: PropTypes.bool,
};

export const transformOtherHelpers: TransformFunc<OtherHelpersProps> = (
  props,
  componentName,
  location = "prop",
) => {
  PropTypes.checkPropTypes(
    otherHelpersPropTypes,
    props,
    location,
    componentName,
  );
  const {
    className,
    marginless,
    paddingless,
    radiusless,
    shadowless,
    unselectable,
    ...rest
  } = props;

  return {
    className: classNames(
      {
        "is-marginless": marginless,
        "is-paddingless": paddingless,
        "is-radiusless": radiusless,
        "is-shadowless": shadowless,
        "is-unselectable": unselectable,
      },
      className,
    ),
    ...rest,
  };
};

/**
 * Responsive
 */
export const BREAKPOINTS = tuple(
  "mobile",
  "tablet",
  "desktop",
  "widescreen",
  "fullhd",
  "touch",
);
export type Breakpoints = (typeof BREAKPOINTS)[number];

export type LimitiedResponsiveBreakpointProps = Partial<{
  display: {
    value: Displays;
  };
  hide: {
    value: boolean;
  };
  textAlignment: {
    value: TextAlignments;
  };
  textSize: {
    value: TextSizes;
  };
}>;

export type ResponsiveBreakpointProps = Partial<{
  display: {
    only?: boolean;
    value: Displays;
  };
  hide: {
    only?: boolean;
    value: boolean;
  };
  textAlignment: {
    only?: boolean;
    value: TextAlignments;
  };
  textSize: {
    only?: boolean;
    value: TextSizes;
  };
}>;

export type ResponsiveHelpersProps = Partial<{
  responsive: Partial<{
    mobile: LimitiedResponsiveBreakpointProps;
    tablet: ResponsiveBreakpointProps;
    desktop: ResponsiveBreakpointProps;
    widescreen: ResponsiveBreakpointProps;
    fullhd: LimitiedResponsiveBreakpointProps;
    touch: LimitiedResponsiveBreakpointProps;
  }>;
}>;

export const responsiveBreakpointPropTypes = {
  display: PropTypes.shape({
    only: PropTypes.bool,
    value: PropTypes.oneOf(DISPLAYS).isRequired,
  }),
  hide: PropTypes.shape({
    only: PropTypes.bool,
    value: PropTypes.bool.isRequired,
  }),
  textAlignment: PropTypes.shape({
    only: PropTypes.bool,
    value: PropTypes.oneOf(TEXT_ALIGNMENTS).isRequired,
  }),
  textSize: PropTypes.shape({
    only: PropTypes.bool,
    value: PropTypes.oneOf(TEXT_SIZES).isRequired,
  }),
};

export const limitedResponsiveBreakpointPropTypes = {
  display: PropTypes.shape({
    value: PropTypes.oneOf(DISPLAYS).isRequired,
  }),
  hide: PropTypes.shape({
    value: PropTypes.bool.isRequired,
  }),
  textAlignment: PropTypes.shape({
    value: PropTypes.oneOf(TEXT_ALIGNMENTS).isRequired,
  }),
  textSize: PropTypes.shape({
    value: PropTypes.oneOf(TEXT_SIZES).isRequired,
  }),
};

export const responsiveHelpersPropTypes = {
  responsive: PropTypes.shape({
    // tslint:disable:object-literal-sort-keys
    mobile: PropTypes.shape(limitedResponsiveBreakpointPropTypes),
    tablet: PropTypes.shape(responsiveBreakpointPropTypes),
    desktop: PropTypes.shape(responsiveBreakpointPropTypes),
    widescreen: PropTypes.shape(responsiveBreakpointPropTypes),
    fullhd: PropTypes.shape(limitedResponsiveBreakpointPropTypes),
    touch: PropTypes.shape(limitedResponsiveBreakpointPropTypes),
    // tslint:enable:object-literal-sort-keys
  }),
};

export const transformResponsiveHelpers: TransformFunc<
  ResponsiveHelpersProps
> = (props, componentName, location = "prop") => {
  PropTypes.checkPropTypes(
    responsiveHelpersPropTypes,
    props,
    location,
    componentName,
  );
  const { className, responsive, ...rest } = props;

  return {
    className: classNames(
      responsive !== undefined
        ? Object.keys(responsive)
            .filter(breakpoint => responsive[breakpoint] !== undefined)
            .map(breakpoint => {
              const names = {};
              const { display, hide, textAlignment, textSize } = responsive[
                breakpoint
              ] as
                | ResponsiveBreakpointProps
                | LimitiedResponsiveBreakpointProps;
              if (display !== undefined) {
                const value = display.value;
                const only = "only" in display ? display.only === true : false;
                names[
                  `is-${value}-${breakpoint}${only ? "-only" : ""}`
                ] = value;
              }
              if (hide !== undefined) {
                const value = hide.value;
                const only = "only" in hide ? hide.only === true : false;
                names[`is-hidden-${breakpoint}${only ? "-only" : ""}`] = value;
              }
              if (textAlignment !== undefined) {
                const value = textAlignment.value;
                const only =
                  "only" in textAlignment ? textAlignment.only === true : false;
                names[
                  `has-text-${value}-${breakpoint}${only ? "-only" : ""}`
                ] = value;
              }
              if (textSize !== undefined) {
                const value = textSize.value;
                const only =
                  "only" in textSize ? textSize.only === true : false;
                names[
                  `is-size-${value}-${breakpoint}${only ? "-only" : ""}`
                ] = value;
              }

              return names;
            })
            .reduce((acc, cv) => ({ ...acc, ...cv }), {})
        : undefined,
      className,
    ),
    ...rest,
  };
};

/**
 * Union of helpers
 */
// tslint:disable-next-line: no-empty-interface
export interface HelpersPropsOverrides {}

export type HelpersProps = Prefer<
  HelpersPropsOverrides,
  FloatHelpersProps &
    OverflowHelpersProps &
    OverlayHelpersProps &
    TypographyHelpersProps &
    VisibilityHelpersProps &
    OtherHelpersProps &
    ResponsiveHelpersProps & { className?: string }
>;

export const helpersPropTypes = {
  ...floatHelpersPropTypes,
  ...overflowHelpersPropTypes,
  ...overlayHelpersPropTypes,
  ...typographyHelpersPropTypes,
  ...visibilityHelpersPropTypes,
  ...otherHelpersPropTypes,
  ...responsiveHelpersPropTypes,
  className: PropTypes.string,
};

export const combineTransformFunctions = <TTransformProps>(
  ...funcs: TransformFunc<any>[] // tslint:disable-line:no-any
): TransformFunc<TTransformProps> => (
  props,
  componentName,
  location = "prop",
) =>
  // tslint:disable-next-line:no-any
  funcs.reduce((acc, func) => func(acc, componentName, location) as any, props);

export const transformHelpers = combineTransformFunctions<HelpersProps>(
  transformFloatHelpers,
  transformOverflowHelpers,
  transformOverlayHelpers,
  transformTypographyHelpers,
  transformVisibilityHelpers,
  transformOtherHelpers,
  transformResponsiveHelpers,
);

export type TransformFunc<TTransformProps extends {}> = <
  TProps extends TTransformProps & { className?: string }
>(
  props: TProps,
  componentName: string,
  location?: string,
) => Omit<TProps, keyof TTransformProps | "className"> & { className?: string };
