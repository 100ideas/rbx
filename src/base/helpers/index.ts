import { Prefer } from "../../types";
import { TransformFunc } from "./types";

import { FloatHelpersProps, transformFloatHelpers } from "./float";
import { OtherHelpersProps, transformOtherHelpers } from "./other";
import { OverflowHelpersProps, transformOverflowHelpers } from "./overflow";
import { OverlayHelpersProps, transformOverlayHelpers } from "./overlay";
import {
  ResponsiveHelpersProps,
  transformResponsiveHelpers,
} from "./responsive";
import {
  transformTypographyHelpers,
  TypographyHelpersProps,
} from "./typography";
import {
  transformVisibilityHelpers,
  VisibilityHelpersProps,
} from "./visibility";

export { TransformFunc } from "./types";

/**
 * Union of helpers
 */
// tslint:disable-next-line: no-empty-interface
export interface HelpersPropsOverrides {}

export type HelpersProps = Prefer<
  HelpersPropsOverrides,
  FloatHelpersProps &
    OverflowHelpersProps &
    OverlayHelpersProps &
    TypographyHelpersProps &
    VisibilityHelpersProps &
    OtherHelpersProps &
    ResponsiveHelpersProps & { className?: string }
>;

export const combineTransformFunctions = <TTransformProps>(
  ...funcs: TransformFunc<any>[] // tslint:disable-line:no-any
): TransformFunc<TTransformProps> => (
  props,
  componentName,
  location = "prop",
) =>
  // tslint:disable-next-line:no-any
  funcs.reduce((acc, func) => func(acc, componentName, location) as any, props);

export const transformHelpers = combineTransformFunctions<HelpersProps>(
  transformFloatHelpers,
  transformOverflowHelpers,
  transformOverlayHelpers,
  transformTypographyHelpers,
  transformVisibilityHelpers,
  transformOtherHelpers,
  transformResponsiveHelpers,
);
