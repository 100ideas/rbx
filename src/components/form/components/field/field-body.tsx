import { cx } from "emotion";
import React from "react";

import Element from "components/element";
import renderAsExoticComponent from "components/render-as-exotic-component";
import { ModifierProps } from "modifiers";

export type FieldBodyModifierProps = Partial<{
  children: React.ReactNode;
  style: {};
}>;

export type FieldBodyProps = ModifierProps & FieldBodyModifierProps;

const FieldBody = renderAsExoticComponent<FieldBodyProps, "div">(
  ({ children, className, ...props }, ref) => (
    <Element {...props} ref={ref} className={cx("field-body", className, {})}>
      {children}
    </Element>
  ),
  "div",
);
FieldBody.defaultProps = Object.assign(
  {
    children: null,
  },
  FieldBody.defaultProps,
);

export default FieldBody;
