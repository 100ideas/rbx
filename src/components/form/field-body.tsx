import { cx } from "emotion";
import React from "react";

import { Element } from "components/element";
import { renderAsExoticComponent } from "components/render-as-exotic-component";
import { ModifierProps } from "modifiers";

export type FieldBodyModifierProps = Partial<{
  children: React.ReactNode;
  style: React.CSSProperties;
}>;

export type FieldBodyProps = ModifierProps & FieldBodyModifierProps;

export const FieldBody = renderAsExoticComponent<FieldBodyProps, "div">(
  ({ children, className, ...props }, ref) => (
    <Element {...props} ref={ref} className={cx("field-body", className, {})}>
      {children}
    </Element>
  ),
  "div",
);
FieldBody.defaultProps = Object.assign(
  {
    children: null,
  },
  FieldBody.defaultProps,
);
