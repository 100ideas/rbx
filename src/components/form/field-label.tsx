import { cx } from "emotion";
import React from "react";

import { Element, renderAsExoticComponent } from "@/components/element";
import { ModifierProps } from "@/modifiers";

export type FieldLabelModifierProps = Partial<{
  children: React.ReactNode;
  size: "small" | "normal" | "medium" | "large";
  style: React.CSSProperties;
}>;

export type FieldLabelProps = ModifierProps & FieldLabelModifierProps;

export const FieldLabel = renderAsExoticComponent<FieldLabelProps, "div">(
  ({ children, className, size, ...props }, ref) => (
    <Element
      {...props}
      ref={ref}
      className={cx("field-label", className, {
        [`is-${size}`]: size,
      })}
    >
      {children}
    </Element>
  ),
  "div",
);
FieldLabel.defaultProps = Object.assign(
  {
    children: null,
  },
  FieldLabel.defaultProps,
);
