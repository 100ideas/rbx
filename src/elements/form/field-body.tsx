import { cx } from "emotion";
import React from "react";

import { forwardRefAs } from "@/base";
import { ModifierProps, transformModifiers } from "@/modifiers";

export type FieldBodyModifierProps = Partial<{ className: string }>;

export type FieldBodyProps = ModifierProps & FieldBodyModifierProps;

export const FieldBody = forwardRefAs<FieldBodyProps, "div">(
  (props, ref) => {
    const { as, ...rest } = transformModifiers(props);
    rest.className = cx("field-body", rest.className);
    return React.createElement(as!, { ref, ...rest });
  },
  { as: "div" },
);
