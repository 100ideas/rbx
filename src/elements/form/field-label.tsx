import classNames from "classnames";
import PropTypes from "prop-types";
import React from "react";

import { forwardRefAs, Generic } from "../../base";
import { HelpersProps } from "../../base/helpers";
import { Prefer } from "../../types";
import { tuple } from "../../utils";

export const FIELD_LABEL_DEFAULTS = {
  sizes: tuple("small", "normal", "medium", "large"),
};

export interface FieldLabelVariablesOverrides {}

export interface FieldLabelVariablesDefaults {
  sizes: (typeof FIELD_LABEL_DEFAULTS["sizes"])[number];
}

export type FieldLabelVariables = Prefer<
  FieldLabelVariablesOverrides,
  FieldLabelVariablesDefaults
>;

export type FieldLabelModifierProps = Partial<{
  size: FieldLabelVariables["sizes"];
}>;

export type FieldLabelOwnProps = HelpersProps & FieldLabelModifierProps;
export type FieldLabelForwardsProps = { className: string };

export const FieldLabel = forwardRefAs<
  "div",
  FieldLabelOwnProps,
  FieldLabelForwardsProps
>(
  ({ className, size, ...rest }, ref) => (
    <Generic
      className={classNames("field-label", { [`is-${size}`]: size }, className)}
      ref={ref}
      {...rest}
    />
  ),
  { as: "div" },
);

FieldLabel.displayName = "Field.Label";
FieldLabel.propTypes = {
  size: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
};
