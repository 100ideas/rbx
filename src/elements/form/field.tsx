import { cx } from "emotion";
import React from "react";

import { forwardRefAs } from "@/base";
import { ModifierProps, transformModifiers } from "@/modifiers";
import { tuple } from "@/utils";
import { FieldBody } from "./field-body";
import { FieldLabel } from "./field-label";

export const FIELD_ALIGNMENTS = tuple("centered", "right");
export type FieldAlignments = (typeof FIELD_ALIGNMENTS)[number];

export const FIELD_KINDS = tuple("addons", "group");
export type FieldKinds = (typeof FIELD_KINDS)[number];

export type FieldModifierProps = Partial<{
  align: FieldAlignments;
  expanded: boolean;
  horizontal: boolean;
  kind: FieldKinds;
  multiline: boolean;
  narrow: boolean;
}>;

export type FieldProps = ModifierProps & FieldModifierProps;

export const Field = Object.assign(
  forwardRefAs<FieldProps, "div">((props, ref) => {
    const {
      as,
      align,
      expanded,
      multiline,
      horizontal,
      kind,
      narrow,
      ...rest
    } = transformModifiers(props);

    let k = null;
    if (kind === "addons") {
      k = "has-addons";
    } else if (kind === "group") {
      k = "is-grouped";
    }

    rest.className = cx("field", rest.className, {
      [`${k}`]: k,
      [`${k}-${align}`]: k && align,
      [`${k}-multiline`]: k === "is-grouped" && multiline,
      "is-expanded": expanded,
      "is-horizontal": horizontal,
      "is-narrow": narrow,
    });

    return React.createElement(as!, { ref, ...rest });
  }, "div"),
  {
    Body: FieldBody,
    Label: FieldLabel,
  },
);

Field.defaultProps = Object.assign(
  {
    horizontal: false,
    multiline: false,
  },
  Field.defaultProps,
);
