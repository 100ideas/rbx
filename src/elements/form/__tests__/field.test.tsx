import Enzyme from "enzyme";
import React from "react";

import { Field, FIELD_ALIGNMENTS, FIELD_KINDS } from "../field";
import { FieldBody } from "../field-body";
import { FieldLabel } from "../field-label";

import { hasProperties } from "@/__tests__/helpers";

describe("Field component", () => {
  hasProperties(Field, {
    Body: FieldBody,
    Label: FieldLabel,
    defaultProps: { as: "div" },
  });

  it("should render as the default element", () => {
    const wrapper = Enzyme.shallow(<Field />);
    expect(wrapper.is("div")).toBe(true);
  });

  it("should render as a custom component", () => {
    const as = "span";
    const wrapper = Enzyme.shallow(<Field as={as} />);
    expect(wrapper.is(as)).toBe(true);
  });

  it("should forward ref", () => {
    const ref = React.createRef<HTMLDivElement>();
    // Enzyme owns outer ref: https://github.com/airbnb/enzyme/issues/1852
    const wrapper = Enzyme.mount(
      <div>
        <Field ref={ref} />
      </div>,
    );
    try {
      expect(ref.current).toBe(wrapper.find(".field").instance());
    } finally {
      wrapper.unmount();
    }
  });

  it("should have bulma className", () => {
    const wrapper = Enzyme.shallow(<Field />);
    expect(wrapper.hasClass("field")).toBe(true);
  });

  it("should preserve custom className", () => {
    const className = "foo";
    const wrapper = Enzyme.shallow(<Field className={className} />);
    expect(wrapper.hasClass(className)).toBe(true);
  });

  FIELD_KINDS.map(kind =>
    FIELD_ALIGNMENTS.map(align =>
      it(`should be aligned ${kind}-${align}`, () => {
        const wrapper = Enzyme.shallow(<Field align={align} kind={kind} />);
        expect(
          wrapper.hasClass(
            kind === "addons" ? `has-addons-${align}` : `is-grouped-${align}`,
          ),
        ).toBe(true);
      }),
    ),
  );

  FIELD_KINDS.map(kind =>
    it(`should be ${kind}`, () => {
      const wrapper = Enzyme.shallow(<Field kind={kind} />);
      expect(
        wrapper.hasClass(kind === "addons" ? "has-addons" : "is-grouped"),
      ).toBe(true);
    }),
  );

  FIELD_KINDS.map(kind =>
    [false, true].map(multiline =>
      it(`should ${
        kind === "group" && multiline ? "" : "not "
      }be multiline when ${kind} and ${
        multiline ? "" : "not "
      }multiline `, () => {
        const wrapper = Enzyme.shallow(
          <Field multiline={multiline} kind={kind} />,
        );
        expect(wrapper.hasClass("is-grouped-multiline")).toBe(
          kind === "group" && multiline,
        );
      }),
    ),
  );

  [false, true].map(expanded =>
    it(`should ${expanded ? "" : "not "}be expanded`, () => {
      const wrapper = Enzyme.shallow(<Field expanded={expanded} />);
      expect(wrapper.hasClass("is-expanded")).toBe(expanded);
    }),
  );

  [false, true].map(horizontal =>
    it(`should ${horizontal ? "" : "not "}be horizontal`, () => {
      const wrapper = Enzyme.shallow(<Field horizontal={horizontal} />);
      expect(wrapper.hasClass("is-horizontal")).toBe(horizontal);
    }),
  );

  [false, true].map(narrow =>
    it(`should ${narrow ? "" : "not "}be narrow`, () => {
      const wrapper = Enzyme.shallow(<Field narrow={narrow} />);
      expect(wrapper.hasClass("is-narrow")).toBe(narrow);
    }),
  );
});
