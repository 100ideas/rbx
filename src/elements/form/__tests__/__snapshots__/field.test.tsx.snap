// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Field component should Exists 1`] = `
Object {
  "$$typeof": Symbol(react.forward_ref),
  "Body": Object {
    "$$typeof": Symbol(react.forward_ref),
    "defaultProps": Object {
      "as": "div",
    },
    "render": [Function],
  },
  "Label": Object {
    "$$typeof": Symbol(react.forward_ref),
    "defaultProps": Object {
      "as": "div",
    },
    "render": [Function],
  },
  "defaultProps": Object {
    "as": "div",
    "horizontal": false,
    "multiline": false,
  },
  "render": [Function],
}
`;

exports[`Field component should export Label and Body as static properties 1`] = `
Object {
  "$$typeof": Symbol(react.forward_ref),
  "defaultProps": Object {
    "as": "div",
  },
  "render": [Function],
}
`;

exports[`Field component should export Label and Body as static properties 2`] = `
Object {
  "$$typeof": Symbol(react.forward_ref),
  "defaultProps": Object {
    "as": "div",
  },
  "render": [Function],
}
`;

exports[`Field component should have field classname 1`] = `
<div
  className="field"
>
  <p
    className="bd-notification is-success"
  >
    <p>
      Default
    </p>
    <p>
      Container
    </p>
  </p>
</div>
`;

exports[`Field component should have field classname with addons classname 1`] = `
<div
  className="field has-addons"
>
  <p
    className="bd-notification is-success"
  >
    <p>
      Default
    </p>
    <p>
      Container
    </p>
  </p>
</div>
`;

exports[`Field component should have field classname with grouped classname 1`] = `
<div
  className="field is-grouped"
>
  <p
    className="bd-notification is-success"
  >
    <p>
      Default
    </p>
    <p>
      Container
    </p>
  </p>
</div>
`;

exports[`Field component should have field-body classname 1`] = `
<div
  className="field-body"
>
  <p
    className="bd-notification is-success"
  >
    <p>
      Default
    </p>
    <p>
      Container
    </p>
  </p>
</div>
`;

exports[`Field component should have field-label classname 1`] = `
<div
  className="field-label"
>
  <p
    className="bd-notification is-success"
  >
    <p>
      Default
    </p>
    <p>
      Container
    </p>
  </p>
</div>
`;
